lib/regexp/regexp.go

```diff
@@ -0,0 +1,283 @@
+package regexp // import "go.starlark.net/lib/regexp"
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"go.starlark.net/starlark"
+	"go.starlark.net/starlarkstruct"
+)
+
+// Module regexp is a Starlark module of functions related to regular expressions.
+// The module defines the following functions:
+//
+//     compile(pattern) - Compiles a pattern in RE2 syntax (https://github.com/google/re2/wiki/Syntax) to a value of type 'regexp'.
+//                        Each call to compile returns a distinct regexp value.
+//                        A regexp value can be used for matching using its matches, find, find_all and other methods.
+//
+//     regexp.find(src) - Returns a string holding the text of the leftmost match in the given string of the regular expression regexp.
+//                         The result is "" if there is no match or if the pattern successfully matches an empty string.
+//
+//     regexp.find_all(src, max) - Returns a new, mutable list of all successive matches of the regular expression regexp. An empty list indicates
+//                                 no match. If max > 0, at most max strings are returned. If max == 0,
+//                                 an empty list is returned. If max < 0, all strings are returned.
+//                                 The parameter max is optional: by default no limit is applied.
+//
+//     regexp.find_submatches(src) - Returns a new, mutable list of strings holding the text of the leftmost match of the regular expression regexp
+//                                   in the given string and the matches, if any, of its subexpressions. An empty list indicates
+//                                   no match.
+//
+//     regexp.matches(src) - Reports whether the given string contains any match of the regular expression regexp.
+//
+//     regexp.replace_all(src, repl) - Returns a copy of the given string, replacing matches of the regular expression regexp
+//                                     with the replacement string repl. Inside repl, backslash-escaped digits (\1 to \9) can be
+//                                     used to insert text matching corresponding parenthesized group from the pattern.
+//                                     \0 in repl refers to the entire matching text.
+//
+//     regexp.replace_all(src, replFunc) - Returns a copy of the given string in which all matches of the regular expression regexp
+//                                         have been replaced by the return value of the replacement function applied to the
+//                                         matched substring.
+//
+//     regexp.split(src, max) - Returns a new, mutable list of strings between all the matches of the regular expression regexp.
+//                              If max > 0, at most max strings are returned knowing that the last string is
+//                              the unsplit remainder. If max == 0, an empty list is returned. If max < 0,
```

Comment:
Go's `split()` behavior is different for the `max=0` case: It returns nil. (In Python the corresponding 0 behavior is to apply no limit.)

I don't see a benefit to Go's behavior of returning a sentinel rather than an empty sequence. It breaks the guarantee that for all non-negative k, `max=k` ensures the result is a list of at most size `k`.